feat(ui): change font & redo dialog design

samuelmasse opened the pull request:
The font for all the UI was changed to Roboto Regular. Since this font is a Google font, I added a <link/> to the Google CDN in the studio and admin index.html. I also added an @import in the channel-web CSS since it wouldn't work otherwise (probably not the best solution but it works).

The confirm dialog now looks pretty much the same as this : https://zpl.io/VY45nKE

I had to used !important everywhere, because for some reason the blueprintjs CSS is loaded after my custom CSS, so everything got overridden. I'm not sure how to resolve this problem.

It should be noted that the styling for the buttons are scoped so they can't be used anywhere else. Since we'll probably need similar buttons at other places in the UI, I'm wondering what would be the best way to share the styling.

If we want to use blueprint components without having to add extra classes and all, we could load a custom-blueprint.css or something to override blueprint styles globally.

samuelmasse committed:
use roboto

samuelmasse committed:
dialog redisign

samuelmasse committed:
channel web maybe fix

allardy commented on modules/channel-web/assets/default.css:
```diff
@@ -1,3 +1,5 @@
+@import url('https://fonts.googleapis.com/css2?family=Roboto&display=swap');
```
Please embed them in the package, we try to avoid as much as possible having external dependencies (some clients have firewalls preventing access)

allardy commented on src/bp/ui-admin/public/index.html:
```diff
@@ -1,23 +1,25 @@
 <!DOCTYPE html>
 <html lang="en">
-  <head>
```
Can you check your prettier / tslint config ? Auto-format should keep that tabulation. If I save this file after you, I get a ton of changes that adds the tabulation

allardy commented on src/bp/ui-admin/public/index.html:
```diff
@@ -1,23 +1,25 @@
+  <link rel="shortcut icon" href="%PUBLIC_URL%/favicon.ico" />
+  <link href="assets/custom-theme.css" rel="stylesheet" />
+  <link href="https://fonts.googleapis.com/css?family=Open+Sans:300,400,600" rel="stylesheet" />
+  <link href="https://fonts.googleapis.com/css2?family=Roboto:wght@400;500&display=swap" rel="stylesheet">
```
Please try to find a solution to include these in the package, so we can get rid of external dependencies

allardy reviewed (commented):
I definitely think we should spend more time to overwrite the blueprint css properly, because it would be a pain to have to specify the className property everywhere just to have a custom css. I don't know what the proper way would be, but it would save a lot of time in the future if we don't have to style elements individually.

To load blueprint CSS before your style, try to put the blueprint css import as the first step in index.ts for the studio and admin, since your style will be imported when ui-shared is loaded first.

EFF commented:
I didn't check the code but I'll support @allardy on this, we want to override the blueprint css properly so it can easily be reusable everywhere. Maybe @MaxCloutier  has some insights to share on this

MaxCloutier commented:
@EFF we had a discussion yesterday Sam and I then confirmed with Yann, forgot to post here, I believe it has yet to be implemented, but we will fork the css theme from blueprint and make changes there so we don't have to overwrite everything

allardy commented:
@MaxCloutier Yes we discussed about it, but the last thing I want to see is having to update all components over the codebase to add custom className properties to apply custom styling

MaxCloutier commented:
@allardy yep that's what I had in mind, but it's not implemented yet in this PR as the current changes sparked that discussion. @samuelmasse correct me if I'm wrong

samuelmasse committed:
local fonts

samuelmasse commented:
I added the woff2 files for Roboto so it doesn't require the Google CDN anymore (ui-admin was already using Google CDN for some reason so I removed that). I had to add the files at three different places (admin, studio, channel-web). Maybe ui-shared could help reduce the amount of duplicated content with a `public `directory.

I'm not sure why channel-web behaves so strangely, it doesn't use the font unless it's loaded exactly in it's default.css file.  I noticed that the css for channel-web is loaded in the ui-studio index.html file. I'm not sure understand why that is.

Also I found multiple woff, woff2, eot and ttf files hanging around. I'm not sure where they are used, I suspect they could probably be deleted. In particular ui-studio has lots of fonts file for the Lato font, which I don't think we'll be using anymore.

Something to note is that the new font (Roboto) has no support for Arabic, Chinese or Japanese. It only supports Latin, Cyrillic, Greek and Vietnamese. When testing with Arabic it fallbacks on another default browser font so it's usable but it doesn't have the same style.

samuelmasse committed:
better placement of css import

samuelmasse committed:
load cutom

allardy commented on src/bp/ui-studio/src/web/index.html:
```diff
@@ -1,25 +1,28 @@
+	<div id="app"></div>
+</body>
+
+</html>
```
Please validate your vscode configuration, this indentation is definitely not right. It should be two spaces, and body should be indented

samuelmasse committed:
blueprint format

samuelmasse committed:
blueprint format correct version

samuelmasse committed:
styling with custom css

samuelmasse committed:
prettier fix

allardy reviewed (dismissed):
Excellent job ! The implementation is clean and future proof :)

The disabled state doesn't seems to be handled correctly (color switch to blue while disabled): 
![image](https://user-images.githubusercontent.com/42552874/78834000-d9508700-79bb-11ea-90f4-4d4507f4573c.png)

MaxCloutier commented on src/bp/ui-shared/src/BaseDialog/index.tsx:
```diff
@@ -5,7 +5,7 @@ import React, { FC } from 'react'
 
 export const BaseDialog: FC<BaseDialogProps> = props => {
-  let width = 500
+  let width = 360
```
I'm worried this might break a lot of existing dialogs, could you add a 'sm' size condition below instead for 360?

samuelmasse commented:
@allardy Ok I'll do that. You can see what properties were changed in the `styling with custom css` commit. Also I'm not sure if this is the best solution. An 11k line file is a pain to work with (at least disabling automatic formatting is mandatory for the file as it takes 8-10 sec). 

We could achieve the same results by loading an smaller .css file on top of it that would modifiy only some aspects of the styling. There would be a lot of `property: unset` but I think the maintainability would be much better.

Another alternative would be to copy the entire .scss source code of blueprint. This one is divided into multiple files, uses variables & everything. It's much more maintainable. If you check the source code of node_modules/@blueprintjs/core/src it might be possible to remove all the .ts files and .md files and only compile the SCSS to create a custom skin. The downside of this is that it might be difficult to understand the source code of blueprint (they really use A LOT of variables).

MaxCloutier commented:
@samuelmasse I didn't know that their scss files were public, but if they are, we probably should use them, it might be a bit hard to understand at first, but it would be much easier if we want to change a primary color or something like that and not be worried to change it everywhere.

allardy commented:
As long as we have a clear history of what was changed in the blueprint styling, because if we need to update and we can't re-apply the styling, it's not gonna be fun :P

MaxCloutier commented:
It would probably be easier to go through their commits and see what changed in their scss files easily as opposed to finding the changes in a compiled css file

MaxCloutier commented:
@samuelmasse are you aware that the styling was applied to more btns than just the one from the confirm dialog?

![Screen Shot 2020-04-08 at 2 49 45 PM](https://user-images.githubusercontent.com/3555826/78837093-37736f00-79a8-11ea-8744-f679ed1b5273.png)

samuelmasse commented:
@MaxCloutier Yep I put the styling on the primary intent. It could be warning or danger intent if we want.

MaxCloutier commented:
@samuelmasse let's put it on warning as it fits the purpose of the confirmDialog better

![Apr-08-2020 15-02-09](https://user-images.githubusercontent.com/3555826/78838006-1875dc80-79aa-11ea-9c04-c37113cede88.gif)
Could you also add a transition on the buttons so the hover effect is smoother and remove the outline when the buttons are in focus? 
`button:focus { transition: background 0.3s; }`
`button:focus { outline: none; }`

samuelmasse committed:
primary -> warning

samuelmasse committed:
sm option for dialog

MaxCloutier commented on src/bp/ui-shared/src/blueprint.css:
```diff
@@ -573,6 +573,11 @@ a > .bp3-code {
   min-width: 30px;
   min-height: 30px;
+  color: #182026;
+  transition: 0.3s;
```
does this work? I thought you'd need to specify either a single property or "all" 🤔

samuelmasse replied:
Yep the transition works. I don't really understand what you mean by this:

> single property or "all"

MaxCloutier replied:
I've always used it with `transition: background 0.3s;` or `transition: all 0.3s;`

MaxCloutier commented on src/bp/ui-shared/src/ConfirmDialog/index.tsx:
```diff
@@ -23,7 +23,7 @@ const ConfirmDialogComponent: FC<ConfirmDialogProps> = props => {
 
   return (
-    <BaseDialog icon="warning-sign" usePortal={false} isOpen onClose={onDecline}>
+    <BaseDialog icon="warning-sign" usePortal={false} isOpen onClose={onDecline} size={props.size}>
```
I would just hardcode the size to "sm" since the confirmDialogs are meant to be as lean and fewer options as possible when adding them somewhere (basically the equivalent of window.confirm). Don't forget to remove the size option you added in the dialogs 😬 

```suggestion
    <BaseDialog icon="warning-sign" usePortal={false} isOpen onClose={onDecline} size="sm">    
```

samuelmasse committed:
confirm dialog always sm

samuelmasse committed:
blueprint scss

samuelmasse committed:
custom scss

samuelmasse commented:
I integrated all the original scss from blueprint. The scss is compiled from main.scss to dist/blueprint.css. 

There seems to be lots of configuration variables already so I'm not sure if it's necessary to include all the code. It would also be possible to modify the scss by importing ~@blueprints/etc... from nodes_module like I showed with my other demo.

The other possibility we mentioned of modifying the scss code directly might also be a better option. Otherwise we get a pretty big time overhead to any small modification of the original scss.

allardy commented:
There are too many changed file here, please open a PR from master to add the vanilla blueprint styling to ui-shared with build steps. Then you can just change the corresponding files in this PR, and it will be much smaller
